### ISSUE TITLE
[i-first] Change kernel meta-data from "iterates_over" to "operates_on"

### ISSUE THREAD

**Issue by arporter** [stfc/PSyclone#866]
For phase I of i-first support for LFRic we need support for a new type of kernel. This kernel (which maps from LFRic to UM Physics kernels) will need to be supplied with entire LFRic fields - as opposed to being called from within a loop over columns or dofs.

**Comment by arporter** [comment]
Currently, LFRic kernel metadata has the "iterates_over" entry. Given that this would then be "none" or equivalent, it seems better (and more consistent with the rest of the kernel metadata) to change this to describe what the kernel does rather than what the PSy layer does.

The proposed changes are:

| iterates_over | operates_on | Interpretation |
|-------------------|------------------|-------------------|
| cells              | cell_column  | Kernel expects to be passed a single column of cells     |
| dofs               | dof               | Kernel expects to be passed data for a single dof           |
|  -                   | domain         | Kernel expects to be passed data for the whole domain |

**Comment by arporter** [comment]
As usual, I'll write the documentation first in case that throws up anything we've not thought of. I'll then send it round to check that everyone is happy.

**Comment by arporter** [comment]
![image](https://user-images.githubusercontent.com/17409212/90491165-77f1af00-e137-11ea-9e78-050461a651b5.png)

**Comment by arporter** [comment]
The code changes thus far have been in the API-agnostic `parse/kernel.py`. To go further I need to work on `dynamo0p3.py` and that will probably clash with the work that @TeranIvy is doing in #823. I shall therefore pause this work.

**Comment by TeranIvy** [comment]
I was just about to writw what I am doing in #823, @arporter, to make sure we are on the same page. One of the review suggestions you had was to define `USER_SUPPLIED_KERNEL_ITERATION_SPACES` in `dynamo0p3.py` and another to put `["cells", "dofs"]` in a constant list. So, my strategy in #823 that I am working on now is:
- Define `BUILTIN_ITERATION_SPACES = ["dofs"]` as a module constant in `dynamo0p3_builtins.py`;
- Define `USER_KERNEL_ITERATION_SPACES = ["cells"]` as a module constant in `dynamo0p3.py`; This list can be later extended with iteration spaces required for Physics kernels;
- Define `VALID_ITERATION_SPACES = USER_KERNEL_ITERATION_SPACES + BUILTIN_ITERATION_SPACES` as a module constant in `dynamo0p3.py` (and this can be imported to `LFRicArgDescriptor` for field access checks).

`VALID_ITERATION_SPACES` can be then used for a map between `iterates_over` and `operates_on`, e.g. 
```
ITERATION_SPACES_MAP = dict(zip(VALID_ITERATION_SPACES, ["cell_column", "dof"]))
```
or `VALID_ITERATION_SPACES` can itself be converted into a dictionary map in this issue. What do you think?

**Comment by arporter** [comment]
That sounds great @TeranIvy and don't worry, I shall wait until #823 is done (or close to) before doing any more here.

**Comment by TeranIvy** [comment]
@arporter and @rupertford , I'll have to change my tactics for the lists of iteration spaces as there are circular imports: `dynamo0p3` imports from `LFRicArgDescriptor` and vice-versa. Also, if `BUILTIN_ITERATION_SPACES` is defined in `dynamo0p3` there is a circular import with `dynamo0p3_builtins`. I suspect there could be more as we move classes from `dynamo0p3` into separate modules in `domain/lfric`.

I think that the neatest solution to avoid circular imports would be to define a separate module in `domain/lfric` for all unallocated `dynamo0p3` constants (i.e. the ones that do not belong to a specific class module already in `domain/lfric`), e.g. `domain/lfric/api_constants.py`. I am happy to do this in #823 for only the iteration space lists (as they are new) and open a new issue to move the rest of `dynamo0p3` constants there.

**Comment by arporter** [comment]
Can you judge whether this is just a symptom of the fact that we're still in the process of breaking up the `dynamo0p3.py` behemoth? If it is, then my preferred solution is to make the imports (very) local to where they are used. This way it's normally possible to prevent circular dependencies without rejigging modules.

**Comment by TeranIvy** [comment]
@arporter, I managed to do that for the circular dependency between `dynamo0p3` and `dynamo0p3_builtins`. The circular dependency between `dynamo0p3` and `LFRicArgDescriptor` is much trickier due to the top-level imports in `dynamo0p3`:
```
...
from psyclone.core.access_type import AccessType
from psyclone.domain.lfric import (FunctionSpace, KernCallAccArgList,
                                   KernCallArgList, KernStubArgList,
                                   LFRicArgDescriptor)
from psyclone.psyir.nodes import Loop, Literal, Schedule
...
```
`LFRicArgDescriptor` constants are used in many places in `dynamo0p3` so localised imports for all of them do not seem as a good solution to me. Another way would be to do something like this
```
...
from psyclone.core.access_type import AccessType
from psyclone.dynamo0p3_builtins import BUILTIN_ITERATION_SPACES
USER_KERNEL_ITERATION_SPACES = ["cells"]
VALID_ITERATION_SPACES = USER_KERNEL_ITERATION_SPACES + BUILTIN_ITERATION_SPACES
from psyclone.domain.lfric import (FunctionSpace, KernCallAccArgList,
                                   KernCallArgList, KernStubArgList,
                                   LFRicArgDescriptor)
from psyclone.psyir.nodes import Loop, Literal, Schedule
...
```
and that is very non-Pythonic.

Additional arguments in favour of a separate place(holder) for constants are: 
- Not all `dynamo0p3` module constants will have a "natural" place inside a class,
- Some of the settings/knowledge in the current constant lists will ultimately be read from a configuration file so we will need a module that processes them - the question is just where and what we do in the meantime.

**Comment by arporter** [comment]
It's the imports of dynamo0p3.py into LFRicArgDescriptor that sound like they need attention to me. I would have expected the dependency to be the other way around (because LFRicArgDescriptor is quite specific but dynamo0p3 is more general). Is there anyway of limiting their scope?

**Comment by TeranIvy** [comment]
It's not so much about the scope as it's about order of imports and the design. `dynamo0p3` needs to import specific information from `LFRicArgDescriptor` to be initialised and that's fine as you say. The problem is other way around as the scope can't really be limited due to imports and creation of lists being at top of the module.

One way around would be to define `USER_KERNEL_ITERATION_SPACES` in `LFRicArgDescriptor` but that doesn't sound as the right place for it - it's a general LFRic API constant.

**Comment by TeranIvy** [comment]
Another alternative is to decide which class the iteration_space list belongs to and move/restructure that class. Then it could be imported to both `dynamo0p3` and `LFRicArgDescriptor`.

I am afraid we need to make at least temporary decision about the design of iteration_space properties as there will be more than just "cells" and "dofs" and they will likely be used for construction and checks in more than one LFRic API class. This is why I suggested a separate module in `domain/lfric` for constants until we finish restructuring of `dynamo0p3` and decide what is read from configuration and what is hard-coded.

**Comment by arporter** [comment]
A separate module might be the way to go although as far as configuration is concerned, currently we store configuration information in a `Configuration` object. I've had a look at lfric_arg_descriptor.py and can't see any immediate uses of dynamo0p3.py. Could you paste here the circular dependency that you're getting?

**Comment by arporter** [comment]
Actually, the more I think about it, the more I like the idea of `domain/lfric/api_constants.py` but I just want to make sure I fully understand the problem...

**Comment by TeranIvy** [comment]
I found a combination of imports that doesn't return circular errors so I'm proceeding without a new module

**Comment by arporter** [comment]
Unfortunately, `Kern._iterates_over` is set in the `Kern` constructor as:

    self._iterates_over = call.ktype.iterates_over

Therefore, I may be best leaving that property of the kernel metadata unchanged too. (Currently I have properties for both the new, LFRic-specific `operates_on` and the original `iterates_over`.)

**Comment by arporter** [comment]
Once #892 is merged, the next step will be to modify virtually all of the remaining LFRic kernel metadata to use operates_on. (A few tests will be left as is to verify that we still support the old style of metadata. This will remain the case until #870 is done.)

**Comment by TeranIvy** [comment]
Merged #893 for the first part of this work.

**Comment by TeranIvy** [comment]
As noted above, closed via #902.
